### PR TITLE
qa: add test_cephfs_mirror_remote_snap_corrupt_fails_synced_snapshot

### PR DESCRIPTION
**Pull request by joscollin**
Fixes: https://tracker.ceph.com/issues/65226

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by joscollin** [comment]
https://pulpito.ceph.com/jcollin-2024-03-30_01:34:47-fs:mirror-wip-jcollin-testing-26032024-distro-default-smithi/

```
2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:ERROR: test_cephfs_mirror_remote_snap_corrupt_fails_synced_snapshot (tasks.cephfs.test_mirroring.TestMirroring)
2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:That making manual changes to the remote .snap directory shows 'peer status' state: "failed"
2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/github.com_joscollin_ceph_618874e9407b153d55bd6296226a5f05d6d07e96/qa/tasks/cephfs/test_mirroring.py", line 1523, in test_cephfs_mirror_remote_snap_corrupt_fails_synced_snapshot
2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:    self.mount_b.run_shell(['mkdir', 'd0/.snap/snap_b'])
2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/github.com_joscollin_ceph_618874e9407b153d55bd6296226a5f05d6d07e96/qa/tasks/cephfs/mount.py", line 786, in run_shell
2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:    return self.client_remote.run(args=args, **kwargs)
2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_e691533f9cbb33d85b2187bba20d7102f098636d/teuthology/orchestra/remote.py", line 523, in run
2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_e691533f9cbb33d85b2187bba20d7102f098636d/teuthology/orchestra/run.py", line 455, in run
2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:    r.wait()
2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_e691533f9cbb33d85b2187bba20d7102f098636d/teuthology/orchestra/run.py", line 161, in wait
2024-03-30T01:59:21.402 INFO:tasks.cephfs_test_runner:    self._raise_for_status()
2024-03-30T01:59:21.402 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_e691533f9cbb33d85b2187bba20d7102f098636d/teuthology/orchestra/run.py", line 181, in _raise_for_status
2024-03-30T01:59:21.402 INFO:tasks.cephfs_test_runner:    raise CommandFailedError(
2024-03-30T01:59:21.402 INFO:tasks.cephfs_test_runner:teuthology.exceptions.CommandFailedError: Command failed on smithi050 with status 1: '(cd /home/ubuntu/cephtest/mnt.1 && exec mkdir d0/.snap/snap_b)'
```

@vshankar 
I was expecting the same^ while writing the test, but just gave it a try. The tests doesn't seem to have permissions to write to the remote .snap directory.

**Comment by joscollin** [comment]
> https://pulpito.ceph.com/jcollin-2024-03-30_01:34:47-fs:mirror-wip-jcollin-testing-26032024-distro-default-smithi/
> 
> ```
> 2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:ERROR: test_cephfs_mirror_remote_snap_corrupt_fails_synced_snapshot (tasks.cephfs.test_mirroring.TestMirroring)
> 2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:That making manual changes to the remote .snap directory shows 'peer status' state: "failed"
> 2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
> 2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
> 2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/github.com_joscollin_ceph_618874e9407b153d55bd6296226a5f05d6d07e96/qa/tasks/cephfs/test_mirroring.py", line 1523, in test_cephfs_mirror_remote_snap_corrupt_fails_synced_snapshot
> 2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:    self.mount_b.run_shell(['mkdir', 'd0/.snap/snap_b'])
> 2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/github.com_joscollin_ceph_618874e9407b153d55bd6296226a5f05d6d07e96/qa/tasks/cephfs/mount.py", line 786, in run_shell
> 2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:    return self.client_remote.run(args=args, **kwargs)
> 2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_e691533f9cbb33d85b2187bba20d7102f098636d/teuthology/orchestra/remote.py", line 523, in run
> 2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
> 2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_e691533f9cbb33d85b2187bba20d7102f098636d/teuthology/orchestra/run.py", line 455, in run
> 2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:    r.wait()
> 2024-03-30T01:59:21.401 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_e691533f9cbb33d85b2187bba20d7102f098636d/teuthology/orchestra/run.py", line 161, in wait
> 2024-03-30T01:59:21.402 INFO:tasks.cephfs_test_runner:    self._raise_for_status()
> 2024-03-30T01:59:21.402 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_e691533f9cbb33d85b2187bba20d7102f098636d/teuthology/orchestra/run.py", line 181, in _raise_for_status
> 2024-03-30T01:59:21.402 INFO:tasks.cephfs_test_runner:    raise CommandFailedError(
> 2024-03-30T01:59:21.402 INFO:tasks.cephfs_test_runner:teuthology.exceptions.CommandFailedError: Command failed on smithi050 with status 1: '(cd /home/ubuntu/cephtest/mnt.1 && exec mkdir d0/.snap/snap_b)'
> ```
> 
> @vshankar I was expecting the same^ while writing the test, but just gave it a try. The tests doesn't seem to have permissions to write to the remote .snap directory.

```
2024-03-30T01:59:00.217 INFO:teuthology.orchestra.run:Running command with timeout 300
2024-03-30T01:59:00.217 DEBUG:teuthology.orchestra.run.smithi050:> (cd /home/ubuntu/cephtest/mnt.1 && exec mkdir d0/.snap/snap_b)
2024-03-30T01:59:00.224 INFO:teuthology.orchestra.run.smithi050.stderr:mkdir: cannot create directory 'd0/.snap/snap_b': Permission denied
2024-03-30T01:59:00.225 DEBUG:teuthology.orchestra.run:got remote process result: 1
```

**Comment by vshankar** [comment]
@joscollin I think the ceph user for the remote fs does not have permissions to create snapshots ('s' caps). We would need to add that for this tests...

**Comment by joscollin** [comment]
> @joscollin I think the ceph user for the remote fs does not have permissions to create snapshots ('s' caps). We would need to add that for this tests...

okay, I'll check that when I'm back to work. As I remember, I couldn't write to remotefs in the vstart without root permissions. So that's by default.

**Comment by joscollin** [comment]
> @joscollin I think the ceph user for the remote fs does not have permissions to create snapshots ('s' caps). We would need to add that for this tests...

@vshankar It already has `rwps`:
```
 17     client.mirror_remote:
 18       - "sudo ceph auth caps client.mirror_remote mon 'allow r' mds 'allow rwps' osd 'allow rw tag cephfs *=*' mgr 'allow r'"
```
Did we miss something?

**Comment by vshankar** [comment]
> > @joscollin I think the ceph user for the remote fs does not have permissions to create snapshots ('s' caps). We would need to add that for this tests...
> 
> @vshankar It already has `rwps`:
> 
> ```
>  17     client.mirror_remote:
>  18       - "sudo ceph auth caps client.mirror_remote mon 'allow r' mds 'allow rwps' osd 'allow rw tag cephfs *=*' mgr 'allow r'"
> ```
> 
> Did we miss something?

What ceph user is running `mkdir d0/.snap/snap_b'` from the test? client.mirror_remote is used by the mirror daemon and not the test case.

**Comment by joscollin** [comment]
> What ceph user is running `mkdir d0/.snap/snap_b'` from the test? client.mirror_remote is used by the mirror daemon and not the test case.

It should be `client.1`. The log/ says `client.4759`

**Comment by joscollin** [comment]
@vshankar Test passed: https://pulpito.ceph.com/jcollin-2024-04-05_06:09:20-fs:mirror-wip-jcollin-testing-26032024-distro-default-smithi/. This is ready for review again

**Comment by vshankar** [comment]
> @vshankar Test passed: https://pulpito.ceph.com/jcollin-2024-04-05_06:09:20-fs:mirror-wip-jcollin-testing-26032024-distro-default-smithi/. This is ready for review again

Nice. Could you explain a bit about how you overcame permission denied when taking a remote snap?

**Comment by vshankar** [comment]
> just prepend a `sudo` before `mkdir` ;D

So the mount already have "s" auth caps 👍

**Comment by joscollin** [comment]
jenkins retest this please

**Comment by joscollin** [comment]
rebased

**Comment by vshankar** [comment]
jenkins test make check

File: qa/tasks/cephfs/test_mirroring.py [hunk 1]
```diff
@@ -1482,3 +1493,45 @@ def test_get_set_mirror_dirty_snap_id(self):
...
+
+        dir_name = f'/{dir_name}'
+        peer_uuid = self.get_peer_uuid(peer_spec)
+        with safe_while(sleep=1, tries=60, action=f'wait for failed status: {peer_spec}') as proceed:
+            while proceed():
+                res = self.mirror_daemon_command(f'peer status for fs: {self.primary_fs_name}',
+                                                 'fs', 'mirror', 'peer', 'status',
+                                                 f'{self.primary_fs_name}@{self.primary_fs_id}', peer_uuid)
+                if('failed' == res[dir_name]['state'] and \
+                   f'{snap_name}' == res[dir_name]['last_synced_snap']['name'] and \
```
Comment (vshankar) [inline qa/tasks/cephfs/test_mirroring.py:1534]: just `snap_name ==` should suffice.

File: qa/tasks/cephfs/test_mirroring.py [hunk 2]
```diff
@@ -1482,3 +1493,45 @@ def test_get_set_mirror_dirty_snap_id(self):
...
+                   expected_snap_count == res[dir_name]['snaps_synced']):
+                    break
+        self.disable_mirroring(self.primary_fs_name, self.primary_fs_id)
```
Comment (vshankar) [inline qa/tasks/cephfs/test_mirroring.py:1537]: At this point, remove the remote snapshot and check if mirroring status recovers and synchronization continues as usual.

**Review by vshankar — CHANGES_REQUESTED** [review]
Otherwise LGTM.

**Comment by joscollin** [comment]
@vshankar Review comments addressed.

**Comment by joscollin** [comment]
@vshankar Test passed: https://pulpito.ceph.com/jcollin-2024-04-10_05:28:15-fs:mirror-wip-jcollin-testing-26032024-distro-default-smithi/

**Comment by joscollin** [comment]
Closing in favour of https://github.com/ceph/ceph/pull/56816

**Pull request closed without merging**